Git update (#82)
kossiitkgp/docs: events/git-and-github.md

Reviewer: Please mention the circulation of an anonymous feedback form. Also, mention some common questions to be added in the form. You can use the previous form as a reference.

    After the demo, show them pictorially all the steps from Forking to Sending a PR. Ask the audience to send a PR. Tell them that sending the PR is the most important takeaway of the workshop. TAs in the workshop can provide individual assistance. If needed, you can show the demo again.

    Finally, merge a PR that contains a well-written description to complete the process. 

12. **Conclusion**

    - Tell them to contribute to beginner-friendly issues at metakgp or KOSS repositories on GitHub. Share the Github Links of both the Github Organizations. Pointing them to some beginner-friendly issues would be great.
    - Tell them about the Github Student pack and its benefits like Domains, Free Servers, and so on.
    - Tell them about Github Pages, and tell them how easily a beautiful blog can be set up within minutes.
    - Inform them about Hacktoberfest.
    - Tell them about starring a repo and the usefulness of starred repos.
    - Tell them about the "watch" feature of a repo, and show them how it can be used to track a project.

>     - Ask them to `star` and `watch` the `events` repo, as important links like slides, and resources will be put here

Reviewer 2: ```suggestion
    - Ask them to `star` and `watch` the [events](https://github.com/kossiitkgp/events) repo, as important links like slides, and resources will be put here
```

    If any swags are available distribute them to the audience.

## After the workshop

Upload the following in the [events](https://github.com/kossiitkgp/events) repo:
- Recording of the workshop, if it's recorded
- Link to the slides

> - Link to extra resources, from which they can revise or learn more advanced stuff. Something like [this](https://github.com/kossiitkgp/events/tree/main/2023/git-github-autumn/resources)

Reviewer 2: ```suggestion
- Link to extra resources, from which they can revise or learn more advanced stuff. Something like [this](https://github.com/kossiitkgp/events/tree/main/2023/Autumn/Git-github/resources)
```

## PLEASE DO NOT

- Introduce an idea or feature directly without explaining why it is needed. Just starting with "Branching is an important feature and to create a new branch, we use `git branch` command" is a bad way to introduce branching. Make sure you firmly establish in the audience's mind why they are learning. If you can't explain why in the beginning make a promise that it will be explained later and keep your promise. Otherwise, our workshops will be no different than most of the boring academic lectures.
- Skip hands-on sessions. Hands-on sessions are the main reason we conduct workshops, else the audience would have simply watched a tutorial video. Repeat demos if required. Provide individual assistance. Make sure the audience gets their hands dirty.
- Skip questions. Git contains many non-intuitive ideas. The audience may be confused while you are explaining the concepts. Re-explain things in a different way if necessary. Encourage the audience to speak out if they are confused. Tell them that Git is non-intuitive and it's okay to ask small doubts.

## KOSS Intro
Include 4-5 slides introducing KOSS, focusing on:
- when we do selections.
- why we do it at that time.

> - what we expect from them.

Reviewer 2: ```suggestion
- what we expect from them.
- what they need to do/learn to get selected
```

- Be able to push and pull code from Github
- Be able to send a PR to an Open Source Repository

## Timeline  
The Workshop should be held around the time when Gymkhana societies have their Introductory Seminar, so that we also show our presence during the same time but with a workshop.

> GymK socities have their Intros during the mid of August, So it's best to conduct the Workshop before the end of August.

Reviewer: ```suggestion
Gymkhana societies have their introductory seminars in mid-August, So it's best to conduct the Workshop before the end of August.
```

- Be able to push and pull code from Github
- Be able to send a PR to an Open Source Repository

## Timeline  

> The Workshop should be held around the time when GymK societies have their Intro-Seminar, so that we also show our presence during the same time but with a Workshop.

Reviewer: ```suggestion
The Workshop should be held around the time when Gymkhana societies have their introductory seminar, so that we also show our presence during the same time but with a workshop.
```

Same for other cases below.

GymK socities have their Intros during the mid of August, So it's best to conduct the Workshop before the end of August.

## Before the workshop

Share an instructions file like [this installation guide](https://github.com/kossiitkgp/events/blob/main/2023/git-github-autumn/installation.md), along with the poster of the workshop. The instructions should contain

- How to install Git for Windows, Linux, Mac seperately
- For Windows users, make that they install `git bash` and configure it
- How to verify that git is installed
- Instructions on how to create a Github Account

In the post mention the fact that we will be available for installation issues before 15 minutes of the workshop. Help with the installation issues before the workshop begins.

## Connecting Github Account
While SSH authentication is a secure and efficient way to interact with GitHub, it may not work on the campus network.  
Alternative :
- Use Personal Access Tokens.
- Use Browser Login Method.

> 

Reviewer: Please remove these extra empty lines

    While demonstrating each and every step explain clearly, why you are doing this. For example, tell them that you are forking the repo to have your own copy of the codebase. Similarly, make your explanation in such a way that steps have a logical sequence.

    After the demo, show them pictorially all the steps from Forking to Sending a PR. Ask the audience to send a PR. Tell them that sending the PR is the most important takeaway of the workshop. TAs in the workshop can provide individual assistance. If needed, you can show the demo again.

    Finally, merge a PR that contains a well-written description to complete the process. 

12. **Conclusion**

    - Tell them to contribute to beginner-friendly issues at metakgp or KOSS repositories on GitHub. Share the Github Links of both the Github Organizations. Pointing them to some beginner-friendly issues would be great.
    - Tell them about the Github Student pack and its benefits like Domains, Free Servers, and so on.
    - Tell them about Github Pages, and tell them how easily a beautiful blog can be set up within minutes.

>     - Inform them about Hacktoberfest and tell them about the T-shirt and Stickers

Reviewer 2: No more t-shirt and sticker :/

    - Tell them about starring a repo and the usefulness of starred repos.
    - Tell them about the "watch" feature of a repo, and show them how it can be used to track a project.
    - Ask them to `star` and `watch` the `events` repo, as important links like slides, and resources will be put here

    If any swags are available distribute them to the audience.

## After the workshop

Upload the following in the `events` repo:
- Recording of the workshop, if it's recorded
- Link to the slides

> - Link to extra resources, from which they can revise or learn more advanced stuff. Something like [this](https://github.com/kossiitkgp/Git-and-Github-Workshop-2023/tree/main/resources)

Reviewer: This link is also outdated. Please go through all links once.

Share the above via a Facebook post along with a thanks to the audience.

## PLEASE DO NOT

- Introduce an idea or feature directly without explaining why it is needed. Just starting with "Branching is an important feature and to create a new branch, we use `git branch` command" is a bad way to introduce branching. Make sure you firmly establish in the audience's mind why they are learning. If you can't explain why in the beginning make a promise that it will be explained later and keep your promise. Otherwise, our workshops will be no different than most of the boring academic lectures.
- Skip hands-on sessions. Hands-on sessions are the main reason we conduct workshops, else the audience would have simply watched a tutorial video. Repeat demos if required. Provide individual assistance. Make sure the audience gets their hands dirty.
- Skip questions. Git contains many non-intuitive ideas. The audience may be confused while you are explaining the concepts. Re-explain things in a different way if necessary. Encourage the audience to speak out if they are confused. Tell them that Git is non-intuitive and it's okay to ask small doubts.

## KOSS Intro

> Include 4-5 slides introducing KOSS, focusing on when we do selections, why we do it at that time, and what we expect from them. This introduction can be inserted in the middle of the workshop when the maximum audience is present.

Reviewer 2: Put all the things to focus on in points

## KOSS Intro
Include 4-5 slides introducing KOSS, focusing on:
- when we do selections.
- why we do it at that time.
- what we expect from them.

This introduction can be inserted in the middle of the workshop when the maximum audience is present.

## Miscellaneous
- We can also consider conducting an Advanced Git workshop, where concepts like `rebase`, `stash` and `bisect` are taught. The duration of this workshop can be shorter than the duration of a regular workshop(3 hours). If such a thing is being planned, make sure to give some git exercises in the gap between git basic workshop and git advanced workshop, so that they are in touch with git concepts.
- `switch` is the new preferred way to switch branches. This might be included in the upcoming workshops. Refer [this](https://www.banterly.net/2021/07/31/new-in-git-switch-and-restore/).

## Feedback Form
Include an _anonymous_ form at the end of the slides and ask the audience to share their experience and suggestions.
Some questions to include in the feedback form can be (Including all the questions below is not necessary):  

> - Can you now explain Git-Github to someone totally new in this field?  (1-5)

Reviewer: ```suggestion
- How well can you now explain Git-Github to someone totally new in this field?  (1-5)
```

Since this is a 1-5 question.

> - Can you revert a commit with errors now? (1-5)

Reviewer: ```suggestion
- Can you revert a commit with errors now? (Yes / No)
```

> - I don't want the changes made by you recently, which are not yet committed. Can you quickly roll them back? (1-5)

Reviewer: ```suggestion
- I don't want the changes made by you recently, which are not yet committed. Can you quickly roll them back? (Yes / No)
```

- How well did you grasp the concept of Git branching? (1-5)

> - Are you confident enough to push that important file into the GitHub repository? Or try pulling the doc? (1-5)

Reviewer: ```suggestion
- Are you confident enough to push that important file into the GitHub repository? Or try pulling the doc? (Yes / No)
```

- How well did you understand the flow of making a Pull Request? (1-5)

> - Were the workshop materials, such as slides and handouts, clear and helpful in your learning process? (1-5)

Reviewer: ```suggestion
- How helpful were the workshop materials, such as slides and handouts, in your learning process? (1-5)
```

> - Do you feel more confident in collaborating with others on GitHub after attending this workshop? (1-5) 

Reviewer: ```suggestion
- Do you feel more confident in collaborating with others on GitHub after attending this workshop? (Yes / No)
```

- Were there any specific topics or concepts related to Git and GitHub that you expected to be covered in the workshop but were not? (descriptive)
- Did you encounter any challenges or areas of confusion during the workshop? If so, please describe them. (descriptive)

> - Did you find the pace of the workshop appropriate, too fast, or too slow? (descriptive)

Reviewer: ```suggestion
- How did you find the pace of the workshop? (too slow, perfect, too fast)
```

- How well did you grasp the concept of Git branching? (1-5)
- Are you confident enough to push that important file into the GitHub repository? Or try pulling the doc? (1-5)
- How well did you understand the flow of making a Pull Request? (1-5)
- Were the workshop materials, such as slides and handouts, clear and helpful in your learning process? (1-5)
- Do you feel more confident in collaborating with others on GitHub after attending this workshop? (1-5) 
- Were there any specific topics or concepts related to Git and GitHub that you expected to be covered in the workshop but were not? (descriptive)
- Did you encounter any challenges or areas of confusion during the workshop? If so, please describe them. (descriptive)
- Did you find the pace of the workshop appropriate, too fast, or too slow? (descriptive)
- How likely are you to apply the knowledge and skills gained from this workshop in your future projects or work? (1-5)
- What more could we have done to improve your learning experience? (descriptive)
- How relevant and useful did you find the workshop? (1-5)
- Any other suggestions? (descriptive)

> 

Reviewer: Please remove extra empty lines here.

    - Tell them about starring a repo and the usefulness of starred repos.
    - Tell them about "watch" feature of a repo, and show them how it can be used to track a project.
    - Ask them to `star` and `watch` the `events` repo, as important links like slides, resources will be put here

    If any swags are available distriubute them to the audience.

## After the workshop

> Upload the following in the `events` repo:  

Reviewer: ```suggestion
Upload the following in the `events` repo:
```

    - Tell them about starring a repo and the usefulness of starred repos.
    - Tell them about "watch" feature of a repo, and show them how it can be used to track a project.
    - Ask them to `star` and `watch` the `events` repo, as important links like slides, resources will be put here

    If any swags are available distriubute them to the audience.

## After the workshop

> Upload the following in the `events` repo  

Reviewer: ```suggestion
Upload the following in the `events` repo:
```

- Skip hands-on sessions. Hands on sessions are the main reason we conduct workshops, else audience would have simply watched a tutorial video. Repeat demos if required. Provide individual assistance. Make sure audience get their hands dirty.
- Skip questions. Git contains many non-intuitive ideas. Audience may be confused while you are explaining the concepts. Re-explain things in a different way if necessary. Encourage audience to speak out if they have confuse. Tell them that Git is non-intuitive and its okay to ask small doubts.

## KOSS Intro

> This git and github workshop will have a significant part, introducing them about KOSS (4-5 slides worth it material). Focusing on when we do selections, why we do it at that time and what we expect from them - no need to introduce to open source.

Reviewer: ```suggestion
Include 4-5 slides introducing KOSS, focusing on when we do selections, why we do it at that time, and what we expect from them. This introduction can be inserted in the middle of the workshop when the maximum audience is present.
```

## Miscellaneous
- We can also consider to conduct a Advanced Git workshop, where concepts like `rebase`, `stash` and `bisect` are taught. The duration of this workshop can be shorter than duration of a regular workshop(3 hours). If such a thing is being planned, make sure to give some git excerises in the gap between git basic workshop and git advanced workshop, so that they are in touch with git concepts.
- `switch` is the new preferred way to switch branches. This might be included in the upcoming workshops. Refer [this](https://www.banterly.net/2021/07/31/new-in-git-switch-and-restore/).

## Feedback Form

> Include a feedback form at the end of the slides and ask the audience to share their experience and suggestions.

Reviewer: ```suggestion
Include an _anonymous_ feedback form at the end of the slides and ask the audience to share their experience and suggestions.
```

- Can you revert a commit with errors now?
- I don't want the changes made by you recently, which are not yet committed. Can you quickly roll them back?
- How well did you grasp the concept of Git branching?
- Are you condifent enough to push that important file into the github repository? Or try pulling the doc?
- How well did you understand the flow of making a Pull Request?
- Were the workshop materials, such as slides and handouts, clear and helpful in your learning process?
- Do you feel more confident in collaborating with others on GitHub after attending this workshop?
- Were there any specific topics or concepts related to Git and GitHub that you expected to be covered in the workshop but were not?
- Did you encounter any challenges or areas of confusion during the workshop? If so, please describe them.
- Did you find the pace of the workshop appropriate, too fast, or too slow?
- How likely are you to apply the knowledge and skills gained from this workshop in your future projects or work?
- What more could we have done to improve your learning experience?
- How relevant and useful did you find the workshop?

> 

Reviewer: Also, mention that including all of the above questions is not important. Try to keep the form short. Also, add a "Any other suggestions?" question at the end.

Some objective questions (1-10 scale) can also be mentioned.
